python/az/aro/azext_aro/custom.py

```diff
@@ -338,3 +301,92 @@ def get_network_resources(cli_ctx, subnets, vnet):
     resources.update(route_tables)
 
     return resources
+
+
+# service_principal_update manages cluster service principal update
+# 1. If called without parameters it should be best-effort
+# 2. If called with parameters it fails if something is not possible
+# 2.1. If client_id is provided, we expect expect secret to be provided too
+#      If only secret is provided - we are updating secret
+#      In any case we will try to verify and update rbac
+# 2.2. TODO: refresh-cluster-service-principal is provided
+def service_principal_update(cli_ctx, oc, client_id=None, client_secret=None):
+    rp_client_sp = None
+    client_sp = None
+    resources = set()
+
+    # if any of these are set - we expect users to have access to fix rbac so we fail
+    fail = client_id is not None or client_secret is not None
+
+    # update client_id without providing secret is not valid.
+    # this acts as dynamic validator
+    if client_id is not None:
+        if client_id != oc.service_principal_profile.client_id and client_secret is None:
+            raise InvalidArgumentValueError("Must specify --client-id with --client-secret.")
```

Comment:
better for this to live in the validator.  Presumably there's a way of telling in the validator whether you're in a create or update call?

Reply:
Validator is static, It does not have with what to compare. And I would like it to stay like this.
We could just be pragmatic and say, `--client-secret` without `--client-id` is NO GO. Give us both or none. And we just don't try to me smart.
I think I prefer this way, than trying to be smarted kid on the block...